examples/Cassie/cassie_rbt_state_estimator.h

```diff
@@ -10,24 +10,111 @@
+#include "examples/Cassie/cassie_utils.h"
 
 namespace dairlib {
 namespace systems {
 
+using Eigen::Vector3d;
+using Eigen::VectorXd;
+using Eigen::MatrixXd;
+using Eigen::Isometry3d;
+using systems::OutputVector;
+
+using drake::systems::Context;
+using drake::systems::EventStatus;
+using drake::systems::DiscreteValues;
+using drake::systems::DiscreteStateIndex;
+
 /// Translates from a TimestamedVector of cassie torque commands into
 /// a cassie_user_in_t struct for transmission to the real robot.
 class CassieRbtStateEstimator : public drake::systems::LeafSystem<double> {
  public:
-  explicit CassieRbtStateEstimator(
-      const RigidBodyTree<double>&);
+  explicit CassieRbtStateEstimator(const RigidBodyTree<double>&,
+                                   bool is_floating_base);
+  void solveFourbarLinkage(double* left_heel_spring,
+                           double* right_heel_spring,
+                           const VectorXd& q_init) const;
+
  private:
-  void Output(const drake::systems::Context<double>& context,
-      OutputVector<double>* output) const;
+  void AssignImuValueToOutputVector(
+      OutputVector<double>* output, const cassie_out_t& cassie_out) const;
+  void AssignActuationFeedbackToOutputVector(
+      OutputVector<double>* output, const cassie_out_t& cassie_out) const;
+  void AssignNonFloatingBaseStateToOutputVector(
+      OutputVector<double>* output, const cassie_out_t& cassie_out) const;
+  void AssignFloatingBaseStateToOutputVector(
+      OutputVector<double>* output, const VectorXd& state_est) const;
+
+  EventStatus Update(const Context<double>& context,
+                     DiscreteValues<double>* discrete_state) const;
+
+  void CopyStateOut(const Context<double>& context,
+                    OutputVector<double>* output) const;
+
+  void contactEstimation(int* left_contact, int* right_contact,
+    DiscreteValues<double>* discrete_state,
+    const OutputVector<double>& output, const double& dt) const;
 
   const RigidBodyTree<double>& tree_;
-  std::map<std::string, int> positionIndexMap_;
-  std::map<std::string, int> velocityIndexMap_;
-  std::map<std::string, int> actuatorIndexMap_;
+  const bool is_floating_base_;
+
+  std::map<std::string, int> position_index_map_;
+  std::map<std::string, int> velocity_index_map_;
+  std::map<std::string, int> actuator_index_map_;
+
+  // Body indices
+  int left_thigh_ind_ = -1;
+  int right_thigh_ind_ = -1;
+  int left_heel_spring_ind_ = -1;
+  int right_heel_spring_ind_ = -1;
+
+  // Input/output port indices
+  int cassie_out_input_port_;
+  int state_input_port_;
+
+  // State indices
+  DiscreteStateIndex time_idx_;
+
+  DiscreteStateIndex state_idx_;
+  DiscreteStateIndex ekf_X_idx_;
+
+  DiscreteStateIndex previous_velocity_idx_;
+
+  DiscreteStateIndex filtered_residue_double_idx_;
+  DiscreteStateIndex filtered_residue_left_idx_;
+  DiscreteStateIndex filtered_residue_right_idx_;
+
+  DiscreteStateIndex ddq_double_init_idx_;
```

Comment:
Many of these variables are not totally clear by their name. Some comments would be helpful.